test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/WpfTestCase.cs

```diff
@@ -0,0 +1,138 @@
+        public string SkipReason => _testCase.SkipReason;
+        public ITestMethod TestMethod => _testCase.TestMethod;
+        public object[] TestMethodArguments => _testCase.TestMethodArguments;
+        public Dictionary<string, List<string>> Traits => _testCase.Traits;
+        public string UniqueID => _testCase.UniqueID;
+
+        public ISourceInformation SourceInformation
+        {
+            get { return _testCase.SourceInformation; }
+            set { _testCase.SourceInformation = value; }
+        }
+
+        public WpfTestCase(IXunitTestCase testCase)
+        {
+            if (testCase == null)
+            {
+                throw new ArgumentNullException(nameof(testCase));
+            }
+
+            _testCase = testCase;
+        }
+
+        [Obsolete("Called by the deserializer", error: true)]
+        public WpfTestCase() { }
+
+        public Task<RunSummary> RunAsync(IMessageSink diagnosticMessageSink,
+            IMessageBus messageBus,
+            object[] constructorArguments,
+            ExceptionAggregator aggregator,
+            CancellationTokenSource cancellationTokenSource)
+        {
+            var taskCompletionSource = new TaskCompletionSource<RunSummary>();
+            var thread = new Thread(() =>
+                {
+                    try
+                    {
+                        // Set up the SynchronizationContext so that any awaits
+                        // resume on the STA thread as they would in a GUI app.
+                        SynchronizationContext.SetSynchronizationContext(new DispatcherSynchronizationContext());
+
+                        // Start off the test method.
+                        var testCaseTask = _testCase.RunAsync(diagnosticMessageSink, messageBus, constructorArguments, aggregator, cancellationTokenSource);
+
+                        // Arrange to pump messages to execute any async work associated with the test.
+                        var frame = new DispatcherFrame();
+                        Task.Run(async delegate
+                        {
+                            try
+                            {
+                                await testCaseTask;
+                            }
+                            finally
+                            {
+                                // The test case's execution is done. Terminate the message pump.
+                                frame.Continue = false;
+                            }
+                        });
+                        Dispatcher.PushFrame(frame);
+
+                        // Report the result back to the Task we returned earlier.
+                        CopyTaskResultFrom(taskCompletionSource, testCaseTask);
+                    }
+                    catch (Exception e)
+                    {
+                        taskCompletionSource.SetException(e);
+                    }
+                });
+
+            thread.SetApartmentState(ApartmentState.STA);
```

Comment:
Does this work on Windows server build agent running as non-interactive service?